libr/asm/arch/pyc/opcode.c

```diff
@@ -0,0 +1,287 @@
+	{ "v3.7.0a2", opcode_37 },
+	{ "v3.7.0a3", opcode_37 },
+	{ "v3.7.0a4", opcode_37 },
+	{ "v3.7.0b1", opcode_37 },
+	{ "v3.7.0b2", opcode_37 },
+	{ "v3.7.0b3", opcode_37 },
+	{ "v3.7.0b4", opcode_37 },
+	{ "v3.7.0b5", opcode_37 },
+	{ "v3.7.0rc1", opcode_37 },
+	{ "v3.7.1", opcode_37 },
+	{ "v3.7.1rc1", opcode_37 },
+	{ "v3.7.1rc2", opcode_37 },
+	{ "v3.7.2", opcode_37 },
+	{ "v3.7.2rc1", opcode_37 },
+	{ "v3.7.3", opcode_37 },
+	{ "v3.7.3rc1", opcode_37 },
+	{ "v3.7.4", opcode_37 },
+	{ "v3.7.4rc1", opcode_37 },
+	{ "v3.7.4rc2", opcode_37 },
+	{ "v3.7.5", opcode_37 },
+	{ "v3.7.5rc1", opcode_37 },
+	{ "v3.7.6", opcode_37 },
+	{ "v3.7.6rc1", opcode_37 },
+	{ "v3.8.0", opcode_38 },
+	{ "v3.8.0a1", opcode_38 },
+	{ "v3.8.0a2", opcode_38 },
+	{ "v3.8.0a3", opcode_38 },
+	{ "v3.8.0a4", opcode_38 },
+	{ "v3.8.0b1", opcode_38 },
+	{ "v3.8.0b2", opcode_38 },
+	{ "v3.8.0b3", opcode_38 },
+	{ "v3.8.0b4", opcode_38 },
+	{ "v3.8.0rc1", opcode_38 },
+	{ "v3.8.1", opcode_38 },
+	{ "v3.8.1rc1", opcode_38 },
+	{ "v3.9.0a1", opcode_39 },
+	{ "v3.9.0a2", opcode_39 },
+	{ "v3.9.0a3", opcode_39 },
+};
+
+bool pyc_opcodes_equal (pyc_opcodes *op, const char *version) {
+    size_t i;
+	for (i = 0; i < sizeof (version_op) / sizeof (version_opcode); i++) {
+		if (!strcmp (version_op[i].version, version)) {
+			if (version_op[i].opcode_func == (pyc_opcodes *(*)())(op->version_sig))
+				return true;
+		}
+	}
+	return false;
+}
+
+pyc_opcodes *get_opcode_by_version (char *version) {
+    size_t i;
+	for (i = 0; i < sizeof (version_op) / sizeof (version_opcode); i++) {
+		if (!strcmp (version_op[i].version, version)) {
+			return version_op[i].opcode_func ();
+		}
+	}
+	return NULL; // No match version
+}
+
+pyc_opcodes *new_pyc_opcodes () {
+	size_t i, j;
+	pyc_opcodes *ret = R_NEW0 (pyc_opcodes);
+	if (!ret) {
+		return NULL;
+	}
+	ret->have_argument = 90;
+	ret->opcodes = malloc (sizeof (pyc_opcode_object) * 256);
+	if (!ret->opcodes) {
+		R_FREE (ret);
+		return NULL;
+	}
+	for (i = 0; i < 256; i++) {
+		ret->opcodes[i].op_name = r_str_newf ("<%u>", i);
+		if (!ret->opcodes[i].op_name) {
+			for (j = 0; j < i; j++) {
+				free (ret->opcodes[j].op_name);
+			}
+			free (ret->opcodes);
+			R_FREE (ret);
+			return NULL;
+		}
+		ret->opcodes[i].type = 0;
+		ret->opcodes[i].op_code = i;
+		ret->opcodes[i].op_push = 0;
+		ret->opcodes[i].op_pop = 0;
+	}
+
+	ret->opcode_arg_fmt = r_list_new ();
+	return ret;
+}
+
+void free_opcode (pyc_opcodes *opcodes) {
+    size_t i;
+	for (i = 0; i < 256; i++)
+		free (opcodes->opcodes[i].op_name);
+	free (opcodes->opcodes);
+	r_list_free (opcodes->opcode_arg_fmt);
+	R_FREE (opcodes);
+}
+
+void add_arg_fmt (pyc_opcodes *ret, char *op_name, const char *(*formatter) (ut32 oparg)) {
+	pyc_arg_fmt *fmt = R_NEW0 (pyc_arg_fmt);
+	fmt->op_name = op_name;
+	fmt->formatter = formatter;
+	r_list_append (ret->opcode_arg_fmt, fmt);
+}
+
+void (def_op) (struct op_parameter par) {
+	free (par.op_obj[par.op_code].op_name);
```

Comment:
Maybe worth to have a poiner to that reference instead of dereferencing it in every line. Also remove space before ( in function definition

Reply:
I think it's acceptable.